File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0733-1.cpp
class Solution {
public:
    void dfs(int r, int c, int initialColor, int newColor, vector<vector<int>>& image,
            vector<vector<char>> &color){
        int m = image.size();
        int n = image[0].size();
        
        image[r][c] = newColor;
        color[r][c] = 1;

        vector<pair<int, int>> neighs = {{r, c-1}, {r, c+1}, {r-1, c}, {r+1, c}};
        for (auto [nr, nc]: neighs){
            if (nr<0 || nr==m || nc<0 || nc == n) continue;
            if (color[nr][nc] == 0 && image[nr][nc] == initialColor)
                dfs(nr, nc, initialColor, newColor, image, color);
        } // end-for        
    } // end-dfs
    
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        int m = image.size();
        int n = image[0].size();
                
        // Did we visit this slot before?
        vector<vector<char>> color(m, vector<char>(n, 0));

        // Flood fill using DFS
        dfs(sr, sc, image[sr][sc], newColor, image, color);

        return image;
    } // end-floodFill
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_323_Design_And_Analysis_Of_Algorithms/Algorithms/18-DFS/LeetCode/LeetCode-0733-1.py
class Solution:
    def dfs(self, r, c, initialColor, newColor, image, color):
        m = len(image)
        n = len(image[0])
        
        image[r][c] = newColor
        color[r][c] = 1

        neighs = ((r, c-1), (r, c+1), (r-1, c), (r+1, c))
        for nr, nc in neighs:
            if nr<0 or nr==m or nc<0 or nc == n: continue
            if color[nr][nc] == 0 and image[nr][nc] == initialColor:
                self.dfs(nr, nc, initialColor, newColor, image, color)
    
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        m = len(image)
        n = len(image[0])
        color = []
        for i in range(m): color.append([0]*n)
        
        self.dfs(sr, sc, image[sr][sc], newColor, image, color)

        return image
